File: packages/volt-cli/src/Volt.Cli/Sync/Extensions.cs
using Volt.Engine.Format.Body;
using Volt.Engine.Item;
namespace Volt.Cli.Sync;


public enum Access { R, Rw }

public sealed record ExtensionDef(string Ext, Access DefaultAccess);

/// <summary>
/// CLI-side extension registry: maps a workspace filename to its access (rw for source items, r for
/// references). The extension list + access is NOT re-declared here — it is DERIVED from
/// <see cref="ItemKind.FileExtensions"/> (the one canonical table), so a new kind is added in exactly one
/// place. A graphical CFC/SFC body is the same rw .fb/.prg/.fun (a push over it is refused by the bridge on
/// live IDE state, not pre-filtered here).
/// </summary>
public static class Extensions
{
    private static readonly ExtensionDef[] All =
        ItemKind.FileExtensions.Select(x => new ExtensionDef(x.Ext, x.IsWritable ? Access.Rw : Access.R)).ToArray();

    private static readonly Dictionary<string, ExtensionDef> ByExt =
        All.ToDictionary(d => "." + d.Ext, StringComparer.OrdinalIgnoreCase);

    private static ExtensionDef? GetByExt(string ext) => ByExt.TryGetValue(ext, out var d) ? d : null;

    private static ExtensionDef? GetByPath(string relPath)
    {
        var slash = relPath.LastIndexOf('/');
        var baseName = slash >= 0 ? relPath.Substring(slash + 1) : relPath;
        var dot = baseName.LastIndexOf('.');
        return dot < 0 ? null : GetByExt(baseName.Substring(dot));
    }

    /// <summary>The full filename from a workspace path ("POUs/FB_Motor.fb" → "FB_Motor.fb"). Matches the
    /// bridge's wire names (which include extensions).
    ///
    /// <para>There was a FOLDER-MARKER arm here that resolved `POUs/.gitkeep` to the folder name `POUs`.
    /// Nothing has written a `.gitkeep` since folders stopped being items (`Materialize`: "legacy `.gitkeep`
    /// files are only READ back — never produced here"), and what survived was a trap: `IsTrackedPath` said yes
    /// while `IsPushable` said no, so a `.gitkeep` an engineer added by hand showed as an outgoing change every
    /// `volt status` and `volt push` answered "nothing to push", forever. It now falls through to the same
    /// foreign-file refusal every other unsyncable file in `src/` gets, which says so by name.</para></summary>
    public static string? FullNameFromPath(string relPath)
    {
        var slash = relPath.LastIndexOf('/');
        var baseName = slash >= 0 ? relPath.Substring(slash + 1) : relPath;
        var dot = baseName.LastIndexOf('.');
        if (dot < 0) return null;
        if (GetByExt(baseName.Substring(dot)) == null) return null;
        // A DUT is on disk as `.struct`/`.enum`/`.union`/`.alias` but on the wire as `.dut` — the one kind both
        // vendors have. Map it back here, at the single path→name seam, so nothing downstream (push, the version
        // gates, the folder maps) ever sees a name the bridge would not recognize.
        var ext = baseName.Substring(dot + 1);
        return ItemKind.IsDutFileExtension(ext)
            ? baseName.Substring(0, dot + 1) + ItemKind.WireExtFor(ext)
            : baseName;
    }

    /// <summary>The extension definition for a full filename ("PLC_PRG.prg" → { ext:"prg", rw }).</summary>
    public static ExtensionDef? DefFromName(string fullName)
    {
        var dot = fullName.LastIndexOf('.');
        return dot < 0 ? null : GetByExt(fullName.Substring(dot));
    }

    public static bool IsTrackedPath(string relPath)
    {
        if (relPath == ".gitattributes") return true;
        return GetByPath(relPath) != null;
    }

    public static bool IsPushable(string relPath) => GetByPath(relPath)?.DefaultAccess == Access.Rw;
    public static bool IsReadOnly(string relPath) => GetByPath(relPath)?.DefaultAccess == Access.R;

    /// <summary>Normalize EVERY workspace file to LF — the bridge always emits LF, so without this Windows git
    /// (core.autocrlf) round-trips the un-attributed read-only kinds through CRLF and pull/push see spurious drift.</summary>
    public static string GitattributesContent() => "* text=auto eol=lf\n";
}

File: packages/volt-git/src/domain/extensions.ts
/**
 * Extension registry — the single source of truth for every workspace file extension Volt tracks.
 * The bridge sends full filenames (name.ext); the CLI only looks up the extension in this flat table
 * to determine access (rw for source items, r for references).
 */

export type DefaultAccess = "r" | "rw";

export interface ExtensionDef {
	readonly ext: string;
	readonly defaultAccess: DefaultAccess;
}

const EXTENSIONS: readonly ExtensionDef[] = [
	// Writable source is named by KIND (bridge: ItemKind.ExtFor). A POU's editability is NOT the extension —
	// a graphical CFC/SFC body is the same .fb/.prg/.fun (materialized as an `(* @volt-graphical: LANG *)`
	// info comment). These default `rw`; a push over a CFC/SFC body is refused by the bridge on live IDE
	// state (not pre-filtered here). Only reference KINDS are read-only by extension.
	{ ext: "fb", defaultAccess: "rw" },
	{ ext: "prg", defaultAccess: "rw" },
	{ ext: "fun", defaultAccess: "rw" },
	{ ext: "itf", defaultAccess: "rw" },
	{ ext: "struct", defaultAccess: "rw" },
	{ ext: "union", defaultAccess: "rw" },
	{ ext: "enum", defaultAccess: "rw" },
	{ ext: "alias", defaultAccess: "rw" },
	{ ext: "gvl", defaultAccess: "rw" },
	{ ext: "library", defaultAccess: "r" },
	{ ext: "device", defaultAccess: "r" },
	{ ext: "projectinfo", defaultAccess: "r" },
	{ ext: "trace", defaultAccess: "r" },
	{ ext: "recipe", defaultAccess: "r" },
	{ ext: "symbols", defaultAccess: "r" },
	{ ext: "task", defaultAccess: "r" },
	{ ext: "image_pool", defaultAccess: "r" },
	{ ext: "parameter_list", defaultAccess: "r" },
	{ ext: "text_list", defaultAccess: "r" },
	{ ext: "recipe_manager", defaultAccess: "r" },
	{ ext: "visualization_manager", defaultAccess: "r" },
	{ ext: "visualization", defaultAccess: "r" },
	{ ext: "library_manager", defaultAccess: "r" },
	{ ext: "class_diagram", defaultAccess: "r" },
	{ ext: "external_types", defaultAccess: "r" },
	{ ext: "tmc", defaultAccess: "r" },
	{ ext: "", defaultAccess: "r" },
] as const;

export const FOLDER_MARKER = ".gitkeep";

const BY_EXT: ReadonlyMap<string, ExtensionDef> = (() => {
	const m = new Map<string, ExtensionDef>();
	for (const def of EXTENSIONS) {
		if (def.ext.length === 0) continue;
		const extWithDot = `.${def.ext.toLowerCase()}`;
		if (m.has(extWithDot)) throw new Error(`registry: duplicate extension '${extWithDot}'`);
		m.set(extWithDot, def);
	}
	return m;
})();

function getByExt(ext: string): ExtensionDef | undefined {
	return BY_EXT.get(ext.toLowerCase());
}

function getByPath(relPath: string): ExtensionDef | undefined {
	const slash = relPath.lastIndexOf("/");
	const base = slash >= 0 ? relPath.slice(slash + 1) : relPath;
	const dot = base.lastIndexOf(".");
	if (dot < 0) return undefined;
	return getByExt(base.slice(dot));
}

/** The full filename from a workspace path ("POUs/FB_Motor.fb" → "FB_Motor.fb"). Folder markers resolve
 *  to the containing folder name. Used to match the bridge's wire names (which include extensions). */
export function fullNameFromPath(relPath: string): string | undefined {
	const slash = relPath.lastIndexOf("/");
	const base = slash >= 0 ? relPath.slice(slash + 1) : relPath;
	if (base === FOLDER_MARKER) {
		if (slash <= 0) return undefined;
		const beforeSlash = relPath.lastIndexOf("/", slash - 1);
		return relPath.slice(beforeSlash + 1, slash);
	}
	const dot = base.lastIndexOf(".");
	if (dot < 0) return undefined;
	if (getByExt(base.slice(dot).toLowerCase()) === undefined) return undefined;
	return base;
}

/** The extension definition for a full filename ("PLC_PRG.prg" → { ext:"prg", defaultAccess:"rw" }). */
export function defFromName(fullName: string): ExtensionDef | undefined {
	const dot = fullName.lastIndexOf(".");
	if (dot < 0) return undefined;
	return getByExt(fullName.slice(dot));
}

export function isTrackedPath(relPath: string): boolean {
	if (relPath.endsWith(`/${FOLDER_MARKER}`) || relPath === FOLDER_MARKER) return true;
	if (relPath === ".gitattributes") return true;
	return getByPath(relPath) !== undefined;
}

export function isPushable(relPath: string): boolean {
	return getByPath(relPath)?.defaultAccess === "rw";
}

export function isReadOnly(relPath: string): boolean {
	return getByPath(relPath)?.defaultAccess === "r";
}

export function gitattributesContent(): string {
	// Normalize EVERY workspace file to LF. The bridge always emits LF and the whole workspace is text
	// (ST, VG, and the read-only manifests). Without a blanket rule, Windows git (core.autocrlf) round-
	// trips the un-attributed read-only kinds (.library/.task/…) through CRLF, so their committed blob
	// differs from the verbatim-LF `volt/ide` baseline and pull/push see spurious, unpushable drift.
	return "* text=auto eol=lf\n";
}
